official/resnet/resnet_shared.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Functions for running Resnet that are shared across datasets."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import argparse
+import os
+
+import tensorflow as tf
+
+
+def resnet_model_fn(features, labels, mode, params, model_class):
+  """Shared functionality for different resnet model_fns."""
+
+  # Generate a summary node for the images
+  tf.summary.image('images', features, max_outputs=6)
+
+  model = model_class(params['resnet_size'], params['data_format'])
+  logits = model(features, mode == tf.estimator.ModeKeys.TRAIN)
+
+  predictions = {
+      'classes': tf.argmax(logits, axis=1),
+      'probabilities': tf.nn.softmax(logits, name='softmax_tensor')
+  }
+
+  if mode == tf.estimator.ModeKeys.PREDICT:
+    return tf.estimator.EstimatorSpec(mode=mode, predictions=predictions)
+
+  # Calculate loss, which includes softmax cross entropy and L2 regularization.
+  cross_entropy = tf.losses.softmax_cross_entropy(
+      logits=logits, onehot_labels=labels)
+
+  # Create a tensor named cross_entropy for logging purposes.
+  tf.identity(cross_entropy, name='cross_entropy')
+  tf.summary.scalar('cross_entropy', cross_entropy)
+
+  # Add weight decay to the loss.
+  loss = cross_entropy + params['weight_decay'] * tf.add_n(
+      [tf.nn.l2_loss(v) for v in tf.trainable_variables()])
+
+  if mode == tf.estimator.ModeKeys.TRAIN:
+    # Scale the learning rate linearly with the batch size. When the batch size
+    # is 128, the learning rate should be 0.1.
+    initial_learning_rate = 0.1 * params['batch_size'] / params['batch_denom']
```

Comment:
Instead of multiple params - `batch_size`, `batch_denorm`, `epochs` etc., my inclination would be to take a single param: `learning_rate`, which is supposed to be a function that returns the learning rate tensor. So, instead of:

```python
tf.estimator.Estimator(model_fn=..., params={
  'batch_size': FLAGS.batch_size,
  'batch_denom': FLAGS.batch_denom,
   ...
})
```

you'd have:

```python
def learning_rate():
   ...

tf.estimator.Estimator(model_fn=..., params={
  'learning_rate' = learning_rate,
})
```

With that, we more cleanly separate out the learning rate decay logic from the rest of the model, make it easier to experiment with for those that are interested, and make for fewer parameters to understand for those who aren't.

Thoughts?

Reply:
I have no issue with that-- I don't have a great sense of what people would want to play with at this stage, so I trust you on that. That would mean each learning_rate would be defined in the dataset-specific file, which separates it from the rest of the training logic, but if you think finding the function is less burdensome in this case than passing in the params, that WFM.